cachi2/core/package_managers/generic/models.py

```diff
@@ -29,9 +43,47 @@ def no_empty_checksums(cls, value: dict[str, str]) -> dict[str, str]:
             raise ValueError("At least one checksum must be provided.")
         return value
 
+    @model_validator(mode="after")
+    def set_target(self, info: ValidationInfo) -> "LockfileArtifact":
+        """Set the target path if not provided and resolve it into an absolute path."""
+        if not self.target:
+            url_path = urlparse(str(self.download_url)).path
+            self.target = Path(url_path).name
+
+        # needs to have output_dir context in order to be able to resolve the target path
+        # and so that it can be used to check for conflicts with other artifacts
+        if not info.context or "output_dir" not in info.context:
+            raise PackageManagerError(
+                "The `LockfileArtifact` class needs to be called with `output_dir` in the context"
+            )
+        output_dir: RootedPath = info.context["output_dir"]
+        self.target = str(output_dir.join_within_root(self.target).path.resolve())
+
+        return self
+
+    @property
+    def formatted_checksums(self) -> List[ChecksumInfo]:
+        """Return the checksums as a list of ChecksumInfo objects."""
+        return [ChecksumInfo(algo, digest) for algo, digest in self.checksums.items()]
+
 
 class GenericLockfileV1(BaseModel):
     """Defines format of the cachi2 generic lockfile, version 1.0."""
 
     metadata: LockfileMetadata
     artifacts: list[LockfileArtifact]
+
+    @model_validator(mode="after")
+    def no_artifact_conflicts(self) -> "GenericLockfileV1":
+        """Validate that all artifacts have unique targets and download_urls."""
+        urls = Counter(a.download_url for a in self.artifacts)
+        targets = Counter(a.target for a in self.artifacts)
+        duplicate_urls = [str(u) for u, count in urls.most_common() if count > 1]
+        duplicate_targets = [t for t, count in targets.most_common() if count > 1]
+        if duplicate_urls or duplicate_targets:
+            raise ValueError(
+                (f"Duplicate download_urls: {duplicate_urls}\n" if duplicate_urls else "")
```

Comment:
```python
>>> "foo " * len([]) + "bar " * len([1])
'bar '
```
I had it in the original suggestion, but this works too (and is arguably more readable).

Reply:
Good catch! I was thinking about `* (len(urls) > 0)`, but did not type it out in fullness. `if..else` is a better way of expressing the idea anyway.